### PR TITLE
DATA-854 remote geometry functionality

### PR DESCRIPTION
**Pull request by biotinker**
Fix to Remote Geometries.

This PR simplifies many of the data structures used to transmit Frame information, and removes others entirely. All frame-related code has been removed from `config`. Remote geometries are now successfully included in frame transmissions from remote robots.

### CODE REVIEW

**Review by raybjork — COMMENTED** [review]
Looks good, lots of nice simplifcations!  I took a super high level pass and left a few comments.  I think as much as possible we should try to have 1:1 rdk structures that mirror proto messages, with corresponding ToProtobuf and FromProtobuf functions (which this is already doing a lot of but there might be one or two more places where we could benefit from this paradigm)

File: referenceframe/frame_system.go [hunk 1]
```diff
@@ -371,6 +359,56 @@ func (sfs *simpleFrameSystem) composeTransforms(frame Frame, inputMap map[string
 	return q, errAll
 }
 
+// MergeFrameSystems will merge fromFS into toFS with an offset frame given by cfg. If cfg is nil, fromFS
```
Comment (tessavitabile) [inline referenceframe/frame_system.go:362]: How is this function related to the PR? Is this a utility we want to provide for working with remote geometries?

**Review by tessavitabile — COMMENTED** [review]
Nice! A couple questions.

File: referenceframe/frame.go [hunk 2]
```diff
@@ -244,16 +244,25 @@ func (sf *staticFrame) Geometries(input []Input) (*GeometriesInFrame, error) {
 }
 
 func (sf *staticFrame) MarshalJSON() ([]byte, error) {
-	transform, err := spatial.PoseMap(sf.transform)
+	temp := StaticFrameCfg{
```
Comment (biotinker) [inline referenceframe/frame.go:247]: @raybjork

File: referenceframe/frame_system.go (hunk 1 above)
Comment (biotinker) [reply referenceframe/frame_system.go:362]: This function was previously in `config/frame.go`, which has been deleted as part of this work, as everything inside it is duplicated within `referenceframe`.

Since this function operates entirely on `referenceframe` structs, it makes sense to put it here rather than moving it elsewhere inside `config`

Comment (tessavitabile) [reply referenceframe/frame_system.go:362]: Thanks!

File: config/frame_system.go [hunk 3]
```diff
@@ -24,45 +27,89 @@ func (part *FrameSystemPart) ToProtobuf() (*pb.FrameSystemConfig, error) {
...
+	fmt.Println("id", frameConfig.ID)
+	fmt.Println("fsc", fsc.Frame)
+	fmt.Println("part", part.FrameConfig)
+	if len(fsc.Kinematics.AsMap()) > 0 {
+		modelBytes, err := json.Marshal(fsc.Kinematics.AsMap())
+		if err != nil {
+			return nil, err
 		}
-		return nil, err
+		modelFrame, err := referenceframe.UnmarshalModelJSON(modelBytes, fsc.Frame.Name)
```
[the remark this answers is not in the document]
Comment (biotinker) [reply config/frame_system.go:104]: This is what we touched on briefly in the meeting with Eliot; swapping the proto field from `bytes` to a `Struct` comes with over-the-wire communication benefits, but with the downside of needing additional processing at the other end.

Re-interpreting a generic `Struct` as json bytes should be super inexpensive, fortunately.

Since the "official" entry point for parsing these structs is starting with json, that's probably the best way to have this, since we don't want to have multiple parsers for the format.

[the remark this answers is not in the document]
Comment (tessavitabile) [reply config/frame_system.go:104]: Ah, thanks for explaining!

File: config/proto_conversions.go [hunk 4]
```diff
@@ -209,21 +209,25 @@ func ResourceLevelServiceConfigFromProto(proto *pb.ResourceLevelServiceConfig) (
...
+			X: pose.Point().X,
+			Y: pose.Point().Y,
+			Z: pose.Point().Z,
 		},
 	}
 
 	var orientation pb.Orientation
 
-	switch oType := frame.Orientation.(type) {
+	switch oType := pose.Orientation().(type) {
```
Comment (spieswl) [inline config/proto_conversions.go:228]: Nit: Consider assigning pose.Orientation() and checking type and doing casting on the assigned variable?

File: referenceframe/frame.go [hunk 5]
```diff
@@ -400,13 +439,18 @@ func (rf *rotationalFrame) Name() string {
 }
 
 func (rf *rotationalFrame) MarshalJSON() ([]byte, error) {
```
Comment (spieswl) [inline referenceframe/frame.go:441]: Frame name should be updated to `revoluteFrame`

Comment (spieswl) [reply referenceframe/frame.go:441]: Actually, seeing how pervasive both this AND translationalFrame are, I wouldn't expect you to change this just yet, we can probably come back later and clean that up.

File: referenceframe/frame_json.go [hunk 6]
```diff
@@ -1,117 +1,86 @@
...
+}
 
-// UnmarshalFrameJSON deserialized json into a reference referenceframe.
-func UnmarshalFrameJSON(data []byte) (Frame, error) {
-	config := FrameMapConfig{}
-	err := json.Unmarshal(data, &config)
-	if err != nil {
-		return nil, err
-	}
+type JointCfg struct {
```
Comment (spieswl) [inline referenceframe/frame_json.go:17]: Nit: Lint will complain about docs for these, but DATA-790 should bring in comments once you rebase here.

File: referenceframe/model.go [hunk 7]
```diff
@@ -62,6 +64,11 @@ func (m *SimpleModel) ChangeName(name string) {
 	m.name = name
 }
 
+// ChangeName changes the name of this model - necessary for building frame systems.
```
Comment (spieswl) [inline referenceframe/model.go:67]: Nit: Copy/paste artifact

File: referenceframe/model_json.go [hunk 8]
```diff
@@ -13,82 +14,64 @@ import (
...
-		Max      float64                `json:"max"` // in mm or degs
-		Min      float64                `json:"min"` // in mm or degs
-		Geometry spatial.GeometryConfig `json:"geometry"`
-	} `json:"dhParams"`
-	RawFrames []FrameMapConfig `json:"frames"`
+	Name         string           `json:"name"`
+	KinParamType string           `json:"kinematic_param_type"`
+	Links        []LinkCfg       `json:"links"`
+	Joints       []JointCfg      `json:"joints"`
+	DHParams     []DHParamCfg    `json:"dhParams"`
```
[the remark this answers is not in the document]
Comment (spieswl) [reply referenceframe/model_json.go:21]: Yes it is, and it also makes it much easier to do in-line "remarshaling" to a `ModelConfig` from other non-JSON data structures, i.e. reconfiguring a URDF to JSON. Maybe something we want to consider for other nested structures?

File: referenceframe/frame.go (hunk 5 above)
Comment (raybjork) [reply referenceframe/frame.go:441]: Yeah there is some naming mismatch throughout rdk.  We internally call them rotational but they are externally referred to as revolute.  I'm going to make a Product ticket to address this

Comment (raybjork) [reply referenceframe/frame.go:441]: PRODUCT-334 now exists

**Review by tessavitabile — COMMENTED** [review]
Looks good! I'm not going to do a thorough review. I just have a few questions about how this all fits together in order to have the motion planning avoid obstacles specified by remote geometries (which I think is the goal, but correct me if I'm wrong):
- The scope said, "WorldState will be updated such that the Additional Transforms field will contain these Static Frames rather than the PoseInFrames it currently does." Does this require an additional proto change?
- Is a user expected to add the Static Frames from remotes to Additional Transforms, or is that something done by rdk? If it's done by rdk, is that park of this PR?
- Is there additional work required to make motion planning avoid obstacles specified by the Static Frames in Additional Transforms in WorldState?

File: config/proto_conversions.go [hunk 9]
```diff
@@ -209,21 +210,25 @@ func ResourceLevelServiceConfigFromProto(proto *pb.ResourceLevelServiceConfig) (
...
+	}
 	proto := pb.Frame{
 		Parent: frame.Parent,
 		Translation: &pb.Translation{
-			X: frame.Translation.X,
-			Y: frame.Translation.Y,
-			Z: frame.Translation.Z,
+			X: pose.Point().X,
+			Y: pose.Point().Y,
+			Z: pose.Point().Z,
```
Comment (raybjork) [inline config/proto_conversions.go:223]: [nit] would be faster if we store pose.Point() in a variable before creating the proto

File: spatialmath/geometry.go [hunk 10]
```diff
@@ -53,8 +54,8 @@ type GeometryConfig struct {
 	R float64 `json:"r"`
 
 	// define an offset to position the geometry
-	TranslationOffset TranslationConfig `json:"translation"`
-	OrientationOffset OrientationConfig `json:"orientation"`
+	TranslationOffset r3.Vector         `json:"translation"`
```
Comment (raybjork) [inline spatialmath/geometry.go:57]: [nit] this should also be omitempty right?

File: spatialmath/geometry.go [hunk 11]
```diff
@@ -134,3 +135,27 @@ func NewGeometryFromProto(geometry *commonpb.Geometry) (Geometry, error) {
...
+	if sphere := geometry.GetSphere(); sphere != nil {
+		if sphere.RadiusMm == 0 {
+			return NewPointCreator(pose, geometry.Label), nil
+		}
+		return NewSphereCreator(sphere.RadiusMm, pose, geometry.Label)
+	}
+	return nil, ErrGeometryTypeUnsupported
+}
+
+// ToProtobuf converts a GeometryCreator to Protobuf.
```
Comment (raybjork) [inline spatialmath/geometry.go:154]: [nit] its operating on a GeometryConfig

File: config/proto_conversions.go [hunk 12]
```diff
@@ -337,9 +350,26 @@ func FrameConfigFromProto(proto *pb.Frame) (*Frame, error) {
 		default:
 			return nil, errors.New("Orientation type unsupported")
 		}
+		orientCfg, err := spatial.NewOrientationConfig(orient)
```
Comment (raybjork) [inline config/proto_conversions.go:353]: [nit] can assign directly to `frame.Orientation` here `frame.Geometry` on line 369, same type of thing on L264

**Review by raybjork — CHANGES_REQUESTED** [review]
Looked through half of it, going to continue in a bit

File: referenceframe/frame.go [hunk 13]
```diff
@@ -166,6 +166,24 @@ type staticFrame struct {
 	geometryCreator spatial.GeometryCreator
 }
 
+// a tailGeometryStaticFrame is a static frame whose geometry is placed at the end of the frame's transform, rather than at the beginning.
+type tailGeometryStaticFrame struct {
```
Comment (raybjork) [inline referenceframe/frame.go:170]: [q] why is this being introduced?

File: referenceframe/frame.go [hunk 14]
```diff
@@ -244,16 +262,25 @@ func (sf *staticFrame) Geometries(input []Input) (*GeometriesInFrame, error) {
...
-		"name":      sf.name,
-		"transform": transform,
+	temp.Orientation = orientationConfig
+
+	if sf.geometryCreator != nil {
+		geometryConfig, err := spatial.NewGeometryConfig(sf.geometryCreator)
+		if err != nil {
+			return nil, err
+		}
+		temp.Geometry = geometryConfig
```
Comment (raybjork) [inline referenceframe/frame.go:281]: can assign directly to this

File: referenceframe/frame.go [hunk 15]
```diff
@@ -400,13 +439,18 @@ func (rf *rotationalFrame) Name() string {
 }
 
 func (rf *rotationalFrame) MarshalJSON() ([]byte, error) {
-	m := FrameMapConfig{
-		"type":    "rotational",
-		"name":    rf.name,
-		"rotAxis": rf.rotAxis,
-		"limit":   rf.limits,
+	if len(rf.limits) > 1 {
+		return nil, errors.New("cannot marshal revolute frame with >1 DOF, use a Model")
```
Comment (raybjork) [inline referenceframe/frame.go:443]: if we use this error message a couple places lets just make it a constant or function

**Review by raybjork — CHANGES_REQUESTED** [review]
Leaving some more comments, still more to review but I figure theres probably another pass I have to do anyway

File: referenceframe/frame_json.go [hunk 16]
```diff
@@ -19,107 +15,107 @@ const (
...
-		return nil, utils.NewUnexpectedTypeError(name, config["name"])
+// NewLinkConfig constructs a config from a Frame.
+func NewLinkConfig(frame staticFrame) (*LinkConfig, error) {
+	var geom *spatial.GeometryConfig
+	orient, err := spatial.NewOrientationConfig(frame.transform.Orientation())
+	if err != nil {
+		return nil, err
 	}
-
-	switch config["type"] {
```
[the remark this answers is not in the document]
Comment (biotinker) [reply referenceframe/frame_json.go:43]: Totally unnecessary now; the `FrameMapConfig` struct it was operating on no longer exists

File: referenceframe/frame_json.go [hunk 17]
```diff
@@ -19,107 +15,107 @@ const (
...
-func decodePose(config FrameMapConfig) (spatial.Pose, error) {
-	var point r3.Vector
+// ParseConfig converts a LinkConfig into a staticFrame.
+func (cfg *LinkConfig) ParseConfig() (Frame, error) {
+	return cfg.ToStaticFrame(cfg.ID)
+}
 
-	err := mapstructure.Decode(config["point"], &point)
+// ToStaticFrame converts a LinkConfig into a staticFrame with a new name.
+func (cfg *LinkConfig) ToStaticFrame(name string) (Frame, error) {
```
[the remark this answers is not in the document]
Comment (biotinker) [reply referenceframe/frame_json.go:77]: One renames the frame, the other does not. Renames are needed because the actual poses become "_origin" frames for remote components

File: referenceframe/frame.go (hunk 13 above)
Comment (biotinker) [reply referenceframe/frame.go:170]: If a component has a geometry, we put that in the "_origin" frame rather than the actual component frame, because for things like arms, the actual component frame is the arm model itself.

The other option would be to make a new type of frame that wraps frames, but then we'd need to marshal that

File: components/arm/client_test.go [hunk 18]
```diff
@@ -232,21 +230,20 @@ func TestClientModel(t *testing.T) {
...
 		ctx context.Context,
-		additionalTransforms []*referenceframe.PoseInFrame,
+		additionalTransforms []*referenceframe.LinkInFrame,
 	) (framesystemparts.Parts, error) {
-		return framesystemparts.Parts{&config.FrameSystemPart{
-			Name: testArmName,
-			FrameConfig: &config.Frame{
-				Parent: referenceframe.World,
+		return framesystemparts.Parts{&referenceframe.FrameSystemPart{
+			FrameConfig: &referenceframe.LinkInFrame{
```
Comment (biotinker) [inline components/arm/client_test.go:245]: LinkInFrame needs a constructor

File: referenceframe/frame.go [hunk 19]
```diff
@@ -329,13 +355,25 @@ func (pf *translationalFrame) Geometries(input []Input) (*GeometriesInFrame, err
...
-		"type":      "translational",
-		"name":      pf.name,
-		"transAxis": pf.transAxis,
-		"limit":     pf.limits,
+	if len(pf.limits) > 1 {
+		return nil, errors.New("cannot marshal translational frame with >1 DOF, use a mobile2DFrame or a Model")
+	}
+	temp := JointConfig{
+		ID:   pf.name,
+		Type: "prismatic",
```
Comment (raybjork) [inline referenceframe/frame.go:363]: [nit] this has a constant we can use now

File: referenceframe/frame.go [hunk 20]
```diff
@@ -329,13 +355,25 @@ func (pf *translationalFrame) Geometries(input []Input) (*GeometriesInFrame, err
...
+		Axis: spatial.AxisConfig{pf.transAxis.X, pf.transAxis.Y, pf.transAxis.Z},
+		Max:  pf.limits[0].Max,
+		Min:  pf.limits[0].Min,
+	}
+	if pf.geometryCreator != nil {
+		geometryConfig, err := spatial.NewGeometryConfig(pf.geometryCreator)
```
Comment (raybjork) [inline referenceframe/frame.go:369]: [nit] can assign directly here

File: referenceframe/frame.go [hunk 21]
```diff
@@ -400,13 +438,18 @@ func (rf *rotationalFrame) Name() string {
...
-		"name":    rf.name,
-		"rotAxis": rf.rotAxis,
-		"limit":   rf.limits,
+	if len(rf.limits) > 1 {
+		return nil, errors.New("cannot marshal revolute frame with >1 DOF, use a Model")
 	}
-	return json.Marshal(m)
+	temp := JointConfig{
+		ID:   rf.name,
+		Type: "revolute",
```
Comment (raybjork) [inline referenceframe/frame.go:446]: same here, can use the new constant

File: referenceframe/frame_system.go [hunk 22]
```diff
@@ -371,6 +371,117 @@ func (sfs *simpleFrameSystem) composeTransforms(frame Frame, inputMap map[string
...
+	part := &FrameSystemPart{
+		FrameConfig: frameConfig,
+	}
+
+	if len(fsc.Kinematics.AsMap()) > 0 {
+		modelBytes, err := json.Marshal(fsc.Kinematics.AsMap())
+		if err != nil {
+			return nil, err
+		}
+		modelFrame, err := UnmarshalModelJSON(modelBytes, frameConfig.Name())
```
Comment (raybjork) [inline referenceframe/frame_system.go:431]: this can just assign to `part.modelFrame`

File: referenceframe/transformable.go [hunk 23]
```diff
@@ -18,14 +18,51 @@ type Transformable interface {
...
+}
+
+// FrameName returns the name of the frame in which the pose was observed. Needed for Transformable interface.
 func (pF *PoseInFrame) FrameName() string {
-	return pF.frame
+	return pF.parent
+}
+
+// Parent returns the name of the frame in which the pose was observed.
+func (pF *PoseInFrame) Parent() string {
```
Comment (raybjork) [inline referenceframe/transformable.go:59]: [minor] we dont need this, it is the same thing as `FrameName()`

File: referenceframe/transformable.go [hunk 24]
```diff
@@ -38,34 +75,50 @@ func (pF *PoseInFrame) Name() string {
...
 	return &PoseInFrame{
-		frame: frame,
-		pose:  pose,
+		parent: frame,
+		pose:   pose,
 	}
 }
 
 // NewNamedPoseInFrame generates a new PoseInFrame and gives it the specified name.
 func NewNamedPoseInFrame(frame string, pose spatialmath.Pose, name string) *PoseInFrame {
```
Comment (raybjork) [inline referenceframe/transformable.go:98]: [q] is this being used anywhere now that we have `NewLinkInFrame`? Would be good to remove if possible

File: robot/framesystem/parts/framesystem_parts.go [hunk 25]
```diff
@@ -5,40 +5,72 @@ import (
...
 	for i, part := range fsp {
-		tra := part.FrameConfig.Translation
-		ori := &spatialmath.EulerAngles{}
-		if part.FrameConfig.Orientation != nil {
-			ori = part.FrameConfig.Orientation.EulerAngles()
+		pose := part.FrameConfig.Pose()
+		tra := pose.Point()
+		ori := pose.Orientation().EulerAngles()
+		geomString := ""
+		if part.FrameConfig.Geometry() != nil {
```
Comment (raybjork) [inline robot/framesystem/parts/framesystem_parts.go:30]: [minor] this section should live in spatialmath as a method on a `GeometryConfig`

**Review by raybjork — CHANGES_REQUESTED** [review]
Just a few minor things before its good to merge

File: referenceframe/frame_system.go (hunk 22 above)
Comment (edaniels) [reply referenceframe/frame_system.go:431]: You want to check the error first though before possibly assigned an errant value to the struct. I'd keep as is

File: referenceframe/transformable.go (hunk 24 above)
Comment (raybjork) [reply referenceframe/transformable.go:98]: Turns out that yes this can be removed.  Have done so in my PR to your branch

**Review by raybjork — APPROVED** [review]
LGTM!

File: referenceframe/model_json.go [hunk 26]
```diff
@@ -4,154 +4,85 @@ import (
...
-	D        float64                `json:"d"`
-	Alpha    float64                `json:"alpha"`
-	Max      float64                `json:"max"` // in mm or degs
-	Min      float64                `json:"min"` // in mm or degs
-	Geometry spatial.GeometryConfig `json:"geometry"`
+	Name         string          `json:"name"`
+	KinParamType string          `json:"kinematic_param_type,omitempty"`
+	Links        []LinkConfig    `json:"links,omitempty"`
+	Joints       []JointConfig   `json:"joints,omitempty"`
+	DHParams     []DHParamConfig `json:"dhParams,omitempty"`
```
Comment (spieswl) [inline referenceframe/model_json.go:16]: Thank you for this, much cleaner now that we have common structs in a sensible location

File: spatialmath/orientation_json.go [hunk 27]
```diff
@@ -21,11 +21,15 @@ const (
...
 type OrientationConfig struct {
 	Type  OrientationType `json:"type"`
-	Value json.RawMessage `json:"value"`
+	Value json.RawMessage `json:"value,omitempty"`
 }
 
 // NewOrientationConfig encodes the orientation interface to something serializable and human readable.
 func NewOrientationConfig(o Orientation) (*OrientationConfig, error) {
+	if o == nil {
+		o = NewZeroOrientation()
```
Comment (spieswl) [inline spatialmath/orientation_json.go:30]: Nice QOL improvement here

**Review by spieswl — APPROVED** [review]
LGTM, love the overall simplifications and improved usability. One thing that struck me was the overall change from `FrameName` to `Parent` throughout: A small terminology change that has a big implication on what one may be envisioning when working with a frame or link. Nothing to do with that information, just a reminder that naming matters.
